src/main/kotlin/app/CodeLongevity.kt

```diff
@@ -0,0 +1,231 @@
+     */
+    fun printme() {
+        val fd = Date(from.commit.getCommitTime().toLong() * 1000).toLocaleString()
+        val fc = "${from.commit.getName()} '${from.commit.getShortMessage()}'"
+        val tc = "${to.commit.getName()} '${to.commit.getShortMessage()}'"
+        val td = Date(to.commit.getCommitTime().toLong() * 1000).toLocaleString()
+        println("Line '$text' - '${from.file}:${from.line}' added in $fc $fd")
+        println("  last known as '${to.file}:${to.line}' in $tc $td")
+    }
+}
+
+/**
+ * Used to compute age of code lines in the repo.
+ */
+class CodeLongevity(repoPath: String, tailRev: String) {
+    val repo = FileRepository(repoPath)
+    val head: RevCommit = RevWalk(repo).parseCommit(repo.resolve("HEAD"));
+    val tail: RevCommit? =
+        if (tailRev != "") RevWalk(repo).parseCommit(repo.resolve(tailRev))
+        else null
+
+    var codeLines: MutableList<CodeLine> = mutableListOf()
+
+    init {
+        compute()
+    }
+
+    fun test(email: String) {
+        var sum: Long = 0
+        var total: Long = 0;
+        for (line in codeLines) {
+            val author = line.from.commit.getAuthorIdent()
+            if (author.getEmailAddress() != email) {
+                continue
+            }
+            //println("got a line, age ${line.age}\n ${line.printme()}")
+            sum += line.age
+            total++
+        }
+
+        //println("All lines:")
+        //codeLines.forEach { line -> line.printme() }
+
+        var avg = if (total > 0) sum / total else 0
+        println("avg code line age for <$email> is ${avg} seconds")
+    }
+
+    /**
+     * Computes age of all lines in the repo through all revisions.
+     */
+    private fun compute() {
+        val treeWalk = TreeWalk(repo)
+        treeWalk.setRecursive(true)
+        treeWalk.addTree(head!!.getTree())
+
+        val files: MutableMap<String, ArrayList<RevCommitLine>> = mutableMapOf()
+
+        // Build a map of file names and their code lines.
+        while (treeWalk.next()) {
+            // TODO(alex): skip binary files.
+            val fileName = treeWalk.getPathString()
+            val fileLoader = repo.open(treeWalk.getObjectId(0))
+            val fileText = RawText(fileLoader.getBytes())
+            val linesNum = fileText.size()
+            var lines = ArrayList<RevCommitLine>(linesNum)
+            var idx = 0
+            while (idx < fileText.size()) {
+                lines.add(RevCommitLine(head, fileName, idx))
+                idx++
+            }
+            files.put(fileName, lines)
+        }
+  
+        val df = DiffFormatter(DisabledOutputStream.INSTANCE)
+        df.setRepository(repo)
+        df.setDetectRenames(true)
+
+        val revWalk = RevWalk(repo)
+        revWalk.markStart(head)
+
+        var commit: RevCommit? = head
+        var parentCommit: RevCommit? = revWalk.next()  // proceed to head
+        do {
+            parentCommit = revWalk.next()
```

Comment:
this one and the one above: when you do markStart then you didn't moved the iterator yet. So I call the method twice. Ideas how to improve a comment are welcome :)